### PR TITLE
io/lab: Explain the use of `mmap()` in loading dynamic libraries

### PR DESCRIPTION
**Pull request by Alex-deVis**
Include a section that explores the advantages of loading dynamic libraries using `mmap()`, rather than `read()`.
The main arguments are `demand paging` and convenience.

Closes https://github.com/open-education-hub/operating-systems/issues/138.

### CODE REVIEW

File: content/chapters/io/lab/content/file-mappings.md [hunk 1]
```diff
@@ -54,6 +54,26 @@ mmap(0x7fb313e84000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENY
...
 The 6th argument is the offset within the file from where to start copying.
 
+### What About `read()`
+
+Let's get the elephant out of the room: why use `mmap()` instead of `read()` for dynamic libraries?
+
+The short answer: for efficiency and convenience.
+
+Getting a dynamic library into memory using `read()` might seem similar to `mmap()`-ing it, but it does not take advantage of how these libraries are used.
+Dynamic libraries are **non-writable** areas in the process VAS, that are **arbitrarily accessed**.
```
Comment (teodutu) [inline content/chapters/io/lab/content/file-mappings.md:64]: This is too coarse. They also contain writable sections (like `.data`) that are subject to copy-on-write.

File: content/chapters/io/lab/content/file-mappings.md [hunk 2]
```diff
@@ -54,6 +54,26 @@ mmap(0x7fb313e84000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENY
...
+Reading and storing a library whose content might end up mostly unused does not sound like the right thing to do.
+
+On the other hand, `mmap()` makes use of `demand paging`.
+As a reminder, `demand paging` is a technique where memory is allocated, but each memory page is loaded when needed.
+Furthermore, `mmap()` will check if the library was already loaded into memory, and if so, it will not use additional memory, thus reducing the startup time.
+As you remember from the Compute chapter, processes sharing memory might lead to [`copy-on-write`](https://open-education-hub.github.io/operating-systems/Lab/Compute/copy-on-write), but this is not the case because the pages are read-only.
+
+Even without considering any OS tweaks, using `read()` to load dynamic libraries would require extra work.
```
Comment (teodutu) [inline content/chapters/io/lab/content/file-mappings.md:72]: And would end up creating unnecessary copies of the non-writable sections.

**Review by teodutu — CHANGES_REQUESTED** [review]

File: content/chapters/io/lab/content/file-mappings.md (hunk 2 above)
Comment (Alex-deVis) [reply content/chapters/io/lab/content/file-mappings.md:73]: I added this in previous paragraph, right after discussing `copy-on-write`.

File: content/chapters/io/lab/content/file-mappings.md (hunk 1 above)
Comment (Alex-deVis) [reply content/chapters/io/lab/content/file-mappings.md]: My bad.

File: content/chapters/io/lab/content/file-mappings.md (hunk 2 above)
Comment (teodutu) [reply content/chapters/io/lab/content/file-mappings.md:73]: Yeah, you're right. I somehow skipped it.

**Review by teodutu — APPROVED** [review]

**Comment by teodutu** [comment]
The linter failure seems unrelated to this PR so I'm merging it regardless.

**Pull request merged**